### DISCUSSION TITLE
Assertion `neighbor' failed Should be non-null

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26206] [Q&A General] [answered]
Today, I updated my MOOSE version from 2023.8.24 to 2023.10.29 using conda. After the update, my program started throwing an error:
```
Assertion `neighbor' failed
Should be non-null
```
However, the same input cards worked perfectly fine in the older version of MOOSE (I have saved the old version and tested it again). Based on the debugger information, I identified the corresponding code:
```
_face_info = _mesh.faceInfo(_current_elem, _current_side);
if (!_face_info)
{
  const Elem * const neighbor = _current_elem->neighbor_ptr(_current_side);
  mooseAssert(neighbor, "Should be non-null");
  const auto neighbor_side = neighbor->which_neighbor_am_i(_current_elem);
  _face_info = _mesh.faceInfo(neighbor, neighbor_side);
  mooseAssert(_face_info, "We need to have retrieved something.");
}
```
This code is part of a new material class derived from the existing PiecewiseConstantByBlockMaterial class in MOOSE. The current function where the code resides is entirely from PiecewiseConstantByBlockMaterial, and I haven't made any modifications. According to the information on the MOOSE website, PiecewiseConstantByBlockMaterial is designed for finite volume methods, but I was able to use it without any issues in my finite element code before. How can I resolve this problem?

**Comment by GiudGiud** [comment]
Hello

I wonder if the building of the mesh info got skipped? @lindsayad 

Otherwise this situation should never happen:
elem, side is not the right couple for the mesh-info
there is no neighbor so neighbor,other side is not the right couple
So essentially you have a face of an element in your mesh that does not have a face info.

Are you using mesh refinement? Adaptivity?

Guillaume

**Reply by hityyds** [reply]
No, this is a very simple one-dimensional test:
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 1000
  xmin = 0
  xmax = 66.0053
[]
```
And the only modification I made to PiecewiseConstantByBlockMaterial is that I assigned a std::vector<std::vector<Real>> to each location, instead of its original Real.

**Reply by GiudGiud** [reply]
did you check the original PiecewiseConstantByBlockMaterial now to check for any change WE made in between you modified it and now?

**Reply by hityyds** [reply]
I have checked, and there is no changes.

**Reply by hityyds** [reply]
I I just tested a sample card for PiecewiseConstantByBlockMaterial provided on the MOOSE website, and there were no issues. This further confuses me.

**Reply by hityyds** [reply]
Is it possible that the issue arises because I am using the finite element method for computations? The description of PiecewiseConstantByBlockMaterial mentions that it is designed for finite volume methods.

**Reply by GiudGiud** [reply]
We are expanding support for finite element and functors. We want this object to work regardless.

Can you share your entire input file or not?
On a public git repo is the preferred way. Copy-paste works if need be

**Reply by hityyds** [reply]
Actually, I have some questions about the code of PiecewiseConstantByBlockMaterial as well. It first allocates material information for non-boundary elements and then retrieves information about the neighboring element on the boundary. However, shouldn't elements on the boundary have no neighbors? The more detailed code is as follows:
```
if (!_bnd)
  {
    mooseAssert(_current_elem,
                "We should be on a block which means we should definitely have a current element");
    auto it = _sub_id_to_prop.find(_current_elem->subdomain_id());
    mooseAssert(it != _sub_id_to_prop.end(),
                "Block restriction must match the subdomain names passed in the "
                "subdomain_to_prop_value parameter");
    _prop[_qp] = it->second;

    return;
  }

  if (_current_elem)
  {
    _face_info = _mesh.faceInfo(_current_elem, _current_side);
    if (!_face_info)
    {
      const Elem * const neighbor = _current_elem->neighbor_ptr(_current_side);
      mooseAssert(neighbor, "Should be non-null");
      const auto neighbor_side = neighbor->which_neighbor_am_i(_current_elem);
      _face_info = _mesh.faceInfo(neighbor, neighbor_side);
      mooseAssert(_face_info, "We need to have retrieved something.");
    }
```

**Reply by hityyds** [reply]
My input cards are very simple, consisting of Mesh+Action+Material+Executioner. Due to the custom action we've written, it's challenging to extract useful information directly from the cards. However, I don't have the authority to disclose the specific implementation details of the Action. I am just a student.

**Reply by hityyds** [reply]
Unable to completely provide you with all the information, I feel truly sorry.

**Reply by hityyds** [reply]
Here is the detailed input card information. I hope it really provides any help:
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 1000
  xmin = 0
  xmax = 66.0053
[]

[Sphinx]
  executioner_type = Eigenvalue
  order = FIRST
  family = MONOMIAL
  num_groups = 1
  quadrature_type = Hexagonal_quadrature
  Sn_order = 4
  dim = 1
  reflective_boundary = 'left 2'
  vacuum_boundary = 'right'
[]


[Materials]
  [./sca]
    type = ScatteringCrossSection
    prop_name = s
    num_groups = 1
    materials = '0'
    sigma_s = '0.03'
  [../]
  [./fis]
    type = ScatteringCrossSection
    prop_name = f
    num_groups = 1
    materials = '0'
    sigma_s = '0.0225'
  [../]
  [./abs]
    type = TotalCrossSection
    prop_name = t
    num_groups = 1
    subdomain_names = '0'
    sigma_t = '0.05'
  [../]
[]


[Executioner]
  type = Eigenvalue
  solve_type = 'PJFNK' 
  free_power_iterations = 3
  #nl_abs_tol = 1e-4
  nl_rel_tol = 1e-5
  l_tol = 0.01
[]

[Outputs]
  execute_on = 'timestep_end'
  exodus = true
[]
```

**Reply by GiudGiud** [reply]
We are not going to be able to help without a minimal working example that uses MOOSE only if you wont share your source code.

**Reply by hityyds** [reply]
Dear GiudGiud, I have emailed you at [guillaume.giudicelli@gmail.com](url) with the subject "MOOSE error: Assertion 'neighbor' failed Should be non-null." In the email, I've attached my source code and test input cards. If you could spare some time to help me resolve the issues in the code, I would greatly appreciate it.

**Reply by GiudGiud** [reply]
Hello

Yes I saw.
Is your code on a private github repo? If so sharing it that is preferred. We are not supposed to download zip files from the internet ^^

A Moose-only MWE is still much preferred than me installing your application.

Guillaume

**Reply by grmnptr** [reply]
I am not sure I got all of the details here but at the moment we only build the faceinfo objects if we have finite volume variables. You can call the mesh methods in your material's initial setup (look at flow diode). The previous behavior built it whenever the acessor was called.

**Reply by grmnptr** [reply]
Sorry it is in PressureDrop postprocessor, not in flowdiode.

**Reply by GiudGiud** [reply, marked as answer]
>  we only build the faceinfo objects if we have finite volume variables.

this is actually probably the problem
@hityyds can you add this to your simulation?
```
[AuxVariables]
  [dummy]
    type = MooseVariableFVReal
  []
[]
```

**Reply by hityyds** [reply]
It ran successfully! Thank you so much for your advice. I have uploaded the source code to my private repository and shared it with you and your colleague, @grmnptr (presumably your colleague). If you would like to review the source code, please check my Collaborator invitations.

**Reply by hityyds** [reply]
In order to normally use it in finite element calculations, I have removed the code related to the _face_info section in PiecewiseConstantByBlockMaterial. According to the code comments, _face_info is mainly used on boundary elements to determine if the position is correct and to handle ghosted elements. However, in finite element methods, we shouldn't need these operations, right? In fact, I have successfully completed several tests, and the program performs well.

The original code for the computeQpProperties() function in PiecewiseConstantByBlockMaterial is as follows:
```
template <bool is_ad>
void
PiecewiseConstantByBlockMaterialTempl<is_ad>::computeQpProperties()
{
  if (!_bnd)
  {
    mooseAssert(_current_elem,
                "We should be on a block which means we should definitely have a current element");
    auto it = _sub_id_to_prop.find(_current_elem->subdomain_id());
    mooseAssert(it != _sub_id_to_prop.end(),
                "Block restriction must match the subdomain names passed in the "
                "subdomain_to_prop_value parameter");
    _prop[_qp] = it->second;

    return;
  }

  if (_current_elem)
  {
    _face_info = _mesh.faceInfo(_current_elem, _current_side);
    if (!_face_info)
    {
      const Elem * const neighbor = _current_elem->neighbor_ptr(_current_side);
      mooseAssert(neighbor, "Should be non-null");
      const auto neighbor_side = neighbor->which_neighbor_am_i(_current_elem);
      _face_info = _mesh.faceInfo(neighbor, neighbor_side);
      mooseAssert(_face_info, "We need to have retrieved something.");
    }
    mooseAssert(_current_elem->build_side_ptr(_current_side)->vertex_average() ==
                    _face_info->faceCentroid(),
                "Making sure we're in the right place");
    auto it = _sub_id_to_prop.find(_current_elem->subdomain_id());
    if (it == _sub_id_to_prop.end())
    {
      // We may be a ghosted material
      const bool current_elem_is_fi_elem = _current_elem == &_face_info->elem();
      const Elem * const other_elem_to_try =
          current_elem_is_fi_elem ? _face_info->neighborPtr() : &_face_info->elem();
      mooseAssert(other_elem_to_try, "This should be non-null");
      it = _sub_id_to_prop.find(other_elem_to_try->subdomain_id());
      mooseAssert(it != _sub_id_to_prop.end(),
                  "Block restriction must match the subdomain names passed in the "
                  "subdomain_to_prop_value parameter");
    }
    _prop[_qp] = it->second;
    return;
  }

  mooseAssert(
      _face_info,
      "We must have set a face info object in order for the PiecewiseConstantByBlockMaterial "
      "class to work on faces");

  // We must be off the domain
  auto it = _sub_id_to_prop.find(_face_info->elem().subdomain_id());
  mooseAssert(it != _sub_id_to_prop.end(),
              "Block restriction must match the subdomain names passed in the "
              "subdomain_to_prop_value parameter");
  _prop[_qp] = it->second;
}
```
The modified code is as follows:
```
template <bool is_ad>
void
PiecewiseConstantByBlockMaterialTempl<is_ad>::computeQpProperties()
{
  mooseAssert(_current_elem,
                "We should be on a block which means we should definitely have a current element");
  auto it = _sub_id_to_prop.find(_current_elem->subdomain_id());
  mooseAssert(it != _sub_id_to_prop.end(),
              "Block restriction must match the subdomain names passed in the "
              "subdomain_to_prop_value parameter");
  _prop[_qp] = it->second;

  return;
}
```

**Reply by lindsayad** [reply]
We should fix this. This is a MOOSE bug

**Reply by GiudGiud** [reply]
Should be an easy fix in the `PiecewiseConstantByBlockMaterial` to request face infos if (_bnd)

**Reply by lindsayad** [reply]
I've opened https://github.com/idaholab/moose/issues/26250
